### PR TITLE
[SPARK-28573][SQL] Convert InsertIntoTable(HiveTableRelation) to DataSource inserting for partitioned table

### PR DESCRIPTION
**Pull request by advancedxy**
## What changes were proposed in this pull request?
Datasource table now supports partition tables long ago. This commit adds the ability to translate
the InsertIntoTable(HiveTableRelation) to datasource table insertion.

## How was this patch tested?
Existing tests with some modification

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 1]
```diff
@@ -197,7 +197,9 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
           Some(partitionSchema))
 
         val logicalRelation = cached.getOrElse {
-          val sizeInBytes = relation.stats.sizeInBytes.toLong
+          val defaultSizeInBytes = sparkSession.sessionState.conf.defaultSizeInBytes
```
Comment (advancedxy) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: For write path, the stats for HiveTable is not calculated as InsertIntoTable's table filed is not added in the children.

So `org.apache.spark.sql.hive.DetermineTableStats` skip this table

**Comment by advancedxy** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
I think we should have a config for it.

**Comment by advancedxy** [comment]
> I think we should have a config for it.

I could add a config for this. But is that really necessary?  Giving users the option to fallback to the old behaviour? 
We don't have a config for the read path, the write path should be the same eventually.

**Comment by xuanyuanking** [comment]
The config is not only the option for users, but also a bandage when we found some cases missed in product env.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala [hunk 2]
```diff
@@ -58,7 +58,11 @@ class HiveCommandSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
         |TBLPROPERTIES('prop1Key'="prop1Val", '`prop2Key`'="prop2Val")
       """.stripMargin)
     sql("CREATE TABLE parquet_tab3(col1 int, `col 2` int)")
-    sql("CREATE TABLE parquet_tab4 (price int, qty int) partitioned by (year int, month int)")
+    sql(
+      """
+        |CREATE TABLE parquet_tab4 (price int, qty int) partitioned by (year int, month int)
+        |STORED AS PARQUET
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: is it a necessary change?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: cc @wangyum do you think this is fixable?

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Maybe we can be recalculate stats by:
https://github.com/apache/spark/blob/485ae6d1818e8756a86da38d6aefc8f1dbde49c2/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/HadoopFsRelation.scala#L74-L77

This is test case:
https://github.com/apache/spark/blob/485ae6d1818e8756a86da38d6aefc8f1dbde49c2/sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala#L1488-L1500

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala (hunk 2 above)
Comment (advancedxy) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: It's modified because I randomly chose it to make sure the insert into partitioned table can be safely converted.
However it should already covered in other test cases.

So it's neutral change. I can revert it if you think it's unnecessary.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala [hunk 3]
```diff
@@ -112,6 +112,15 @@ private[spark] object HiveUtils extends Logging {
     .booleanConf
     .createWithDefault(true)
 
+  val CONVERT_INSERTING_PARTITIONED_TABLE =
+    buildConf("spark.sql.hive.convertInsertingPartitionedTable")
+      .doc("When set to true, and \"spark.sql.hive.convertMetastoreParquet\" or " +
+        "\"spark.sql.hive.convertMetastoreOrc\" is true, the built-in ORC/Parquet writer is used" +
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala]: Usually we write like \`spark.sql.hive.convertMetastoreOrc\`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetMetastoreSuite.scala [hunk 4]
```diff
@@ -512,16 +512,14 @@ class HiveParquetMetastoreSuite extends ParquetPartitioningTest {
         |PARTITION (`date`='2015-04-01')
         |select a, b from jt
       """.stripMargin)
-    // Right now, insert into a partitioned Parquet is not supported in data source Parquet.
-    // So, we expect it is not cached.
-    assert(getCachedDataSourceTable(tableIdentifier) === null)
+    checkCached(tableIdentifier)
```
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetMetastoreSuite.scala]: This change also affects ORC. Could you add test for Orc?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 5]
```diff
@@ -209,13 +211,16 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
 
           val updatedTable = inferIfNeeded(relation, options, fileFormat, Option(fileIndex))
 
+          // Spark SQL's data source table now support static and dynamic partition insert. Source
+          // table converted from Hive table should always use dynamic.
+          val enableDynamicPartition = options.updated("partitionOverwriteMode", "dynamic")
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: I'm not sure about this part. Why it should always use dynamic?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala (hunk 3 above)
Comment (advancedxy) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala]: I referred `CONVERT_METASTORE_PARQUET_WITH_SCHEMA_MERGING`, will fix it.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetMetastoreSuite.scala (hunk 4 above)
Comment (advancedxy) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetMetastoreSuite.scala]: I will check existing tests. ORC cases are not that much probably, I will add similar cases for ORC then.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 5 above)
Comment (advancedxy) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Please refer https://issues.apache.org/jira/browse/SPARK-20236.
TL;DR version: dynamic follows hive's style, static is the spark style which is different with hive. When converting InsertIntoHiveTable, we definitely want the hive behaviour. So it takes preference over `spark.sql.sources.partitionOverwriteMode`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala (hunk 2 above)
Comment (advancedxy) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: reverted

**Comment by advancedxy** [comment]
ping @cloud-fan and @viirya.
I pushed a new commit since no more comments are made.

**Comment by cloud-fan** [comment]
Can we update `DetermineTableStats` to catch `InsertIntoTable(HiveTableRelation)` as well? Similar to `ResolveRelations`.

**Comment by advancedxy** [comment]
> Can we update `DetermineTableStats` to catch `InsertIntoTable(HiveTableRelation)` as well? 

will push a new commit later.

**Comment by advancedxy** [comment]
@cloud-fan do you have any other suggestions?

The previous failure looks unrelated.

**Comment by advancedxy** [comment]
Rebased with current master. ping @cloud-fan

**Comment by advancedxy** [comment]
> Test build #109863 has finished for PR 25306 at commit 9899ade.
> 
> This patch fails Spark unit tests.
This patch merges cleanly.
This patch adds no public classes.

These test failures are introduced by this change. https://github.com/apache/spark/blob/9899adea2f125c350792d1d7b7ded3a629970766/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala#L146-L151

Once we enables the `InsertIntoTable` with stats,  the command executed in `InsertIntoHiveTable` as ` CommandUtils.updateTableStats(sparkSession, table)` will actually update stats, and invalid
a lot of assumptions.

Do we really want to calculate stats when inserting into tables?  @cloud-fan

**Comment by cloud-fan** [comment]
@advancedxy this a good point. It looks to me that we mess up 2 things:
1. the table statistics from metastore
2. the data size statistics from HDFS

For table stats, we have special logic for it in places like `CommandUtils.updateTableStats`. However, in `DetermineTableStats`, we get size stats from HDFS and set it as table stats, which is incorrect.

My proposal is
```
class HiveTableRelation(..., statsOpt: Option[Statistics] = None) {
  ...
  override def computeStats(): Statistics = {
    tableMeta.stats.map(_.toPlanStats(output, conf.cboEnabled))
      .orElse(statsOpt)
      .getOrElse {
        throw new IllegalStateException("table stats must be specified.")
      }
  }
}
```

And `DetermineTableStats` should set `HiveTableRelation.statsOpt`

**Comment by advancedxy** [comment]
@cloud-fan your proposal sounds good enough to me.

how would you like that to be addressed? 
1. Addressed in this pr.
2. Addressed in another Jira/pr first, and gathers others' opinion if any.

**Comment by cloud-fan** [comment]
I think we can do it in this PR. Without this pr, the relation stats is useless if the relation is inside `InsertTable`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 6]
```diff
@@ -136,10 +136,9 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
              |)""".stripMargin)
 
         spark.sql("REFRESH TABLE t1")
-        // Before SPARK-19678, sizeInBytes should be equal to dataSize.
-        // After SPARK-19678, sizeInBytes should be equal to DEFAULT_SIZE_IN_BYTES.
+        // After SPARK-28573, sizeInBytes should be equal to dataSize.
         val relation1 = spark.table("t1").queryExecution.analyzed.children.head
-        assert(relation1.stats.sizeInBytes === spark.sessionState.conf.defaultSizeInBytes)
+        assert(relation1.stats.sizeInBytes === dataSize)
```
Comment (advancedxy) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: cc @wangyum, after changes applied to HiveTableRelation, we don't modify `tableMeta` any more. 
The converted `LogicalRelation(HadoopFsRelation, xx)` from `HiveTableRelation` will actual get size by `HadoopFsRelation.sizeInBytes`. 

Do you think this is an acceptable change or we should deal with stats differently in `DetermineTableStats`

**Comment by advancedxy** [comment]
> cc @wangyum, after changes applied to HiveTableRelation, we don't modify tableMeta any more.

Gently ping @wangyum and cc @cloud-fan

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala [hunk 7]
```diff
@@ -228,4 +229,64 @@ class HiveOrcQuerySuite extends OrcQueryTest with TestHiveSingleton {
...
+          withTable("dummy_orc_partitioned") {
+            withTempPath { dir =>
+              spark.sessionState.refreshTable("dummy_orc_partitioned")
+              val path = dir.getCanonicalPath
+              spark.sql(
+                s"""
+                   |CREATE TABLE dummy_orc_partitioned(key INT, value STRING)
+                   |PARTITIONED by (`date` STRING)
+                   |STORED AS ORC
+                   |LOCATION '${dir.toURI}'
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala]: we don't need to specify the table location, and save the `withTempPath`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: looks better to me.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 1 above)
Comment (advancedxy) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Fixed by DetermineTableStats handling InsertIntoTable(HiveTableRelation, ...)

**Comment by cloud-fan** [comment]
thanks, merging to master!
